Pipeline/eval

deep-diver opened the pull request:
this PR basically runs the programs written in a jupyter notebook for model evaluation #5 

when it is run, the following is printed out
```
eval on 0...similarity_score: 12, precision_score: 20
eval on 1...similarity_score: 30, precision_score: 50
eval on 2...similarity_score: 85, precision_score: 90
eval on 3...similarity_score: 50, precision_score: 25
eval on 4...similarity_score: 70, precision_score: 80
eval on 5...similarity_score: 85, precision_score: 95
eval on 6...similarity_score: 90, precision_score: 95
eval on 7...similarity_score: 10, precision_score: 0
eval on 8...similarity_score: 25, precision_score: 50
eval on 9...similarity_score: 30, precision_score: 10
eval on 10...similarity_score: 30, precision_score: 50
eval on 11...similarity_score: 45, precision_score: 20
eval on 12...similarity_score: 20, precision_score: 0
eval on 13...similarity_score: 80, precision_score: 85
eval on 14...similarity_score: 60, precision_score: 85
eval on 15...similarity_score: 10, precision_score: 0
qualification: False, avg_similarity_scores: 45.75, avg_precision_scores: 47.1875
```

and it returns the following values for the downstream jobs
```
qualification: False, avg_similarity_scores: 45.75, avg_precision_scores: 47.1875
```

This code runs the text generation inference with batch size of 1. In the later follow-up PR, text generation with arbitrary batch sizes should be supported. To do this, we need the following code modifications:
- batch generation on a given test dataset with local language model (Gemma)
- simultaneous API call to service language model (Gemini) for the evaluation on each generated output

deep-diver committed:
initial commit for eval pipeline

deep-diver committed:
utility functions to organize input steps

deep-diver committed:
remove default parameter values

deep-diver committed:
basic CLI for eval

deep-diver committed:
fix: missing toml pkg

deep-diver committed:
print out each eval result

deep-diver committed:
branch on valid condition

deep-diver committed:
add gitignore

deep-diver committed:
fix typo: gemini_api_key -> api_key

deep-diver committed:
resulting JSON structure validation check

deep-diver committed:
print out eval step

deep-diver committed:
add docstring for eval

deep-diver committed:
add docstrings

deep-diver committed:
add option for load_in_8bit and 4bit

sayakpaul commented on src/pipeline/eval.py:
```diff
@@ -0,0 +1,83 @@
+    avg_precision_scores = total_precision_scores / len(ds)
+    
+    qualification = avg_similarity_scores >= avg_similarity_threshold and avg_precision_scores >= avg_precision_threshold
+    return qualification, (avg_similarity_scores, avg_precision_scores)
```
Should be wrapped in a dictionary:

```python
{"qualification": qualification, "avg_similarity_scores": avg_similarity_scores, "avg_precision_scores": avg_precision_scores}
```

sayakpaul commented on src/utils.py:
```diff
@@ -0,0 +1,47 @@
+def _sort_step_order(input_steps):
```
What is this for?

deep-diver replied:
I just assumed user could enter the steps in arbitrary order like "eval" "ft" "deploy"... This sort them out in the right order "ft" "eval" "deploy" ...

sayakpaul commented on src/main.py:
```diff
@@ -0,0 +1,54 @@
+        raise ValueError("both load_in_8bit and load_in_4bit are set. "
+                         "only one of them should be set at a time")
+    
+    valid, input_steps = get_steps_right(args.steps)
```
What does `get_steps_right()` mean?

deep-diver replied:
the arg `--steps` can take steps such as ft(fine-tuning), eval(evaluation), synth-gen(synthetic data generation), deploy(deployment). However, we can't assume the input is given in a valid form in terms of order, duplications, ...

`get_steps_right()` makes sure if everything is right in that manner.

sayakpaul replied:
`validate_steps()` might be a better name?

sayakpaul commented on src/main.py:
```diff
@@ -0,0 +1,54 @@
+                         "only one of them should be set at a time")
+    
+    valid, input_steps = get_steps_right(args.steps)
+    if valid:
+        if "ft" in input_steps:
+            pass
```
These can be made more readable.

deep-diver replied:
I left it `pass` for the future implementation. Could you elaborate?

sayakpaul replied:
"ft" is a little too short of a name. Spelling it out fully will be more useful ("fine-tuning").

sayakpaul commented on src/pipeline/eval.py:
```diff
@@ -0,0 +1,83 @@
+    total_similarity_scores = 0
+    total_precision_scores = 0
+    ds = _get_test_dataset(test_dataset_id, test_dataset_split)
+    eval_prompt_tmpl = _get_eval_prompt_tmpl(eval_prompt_tmpl_path)
```
Is it `tmpl` or `toml`?

deep-diver replied:
toml could contain multiple templates, so here `tmpl` is the right term

sayakpaul commented on src/pipeline/eval.py:
```diff
@@ -0,0 +1,83 @@
+    ds = _get_test_dataset(test_dataset_id, test_dataset_split)
+    eval_prompt_tmpl = _get_eval_prompt_tmpl(eval_prompt_tmpl_path)
+    
+    for idx, data in enumerate(ds):
+        assessments = _eval_on_single_record(model, tokenizer, data, eval_prompt_tmpl, gemini_api_key)
+        similarity_score = assessments['similarity_assessment']['score']
+        precision_score = assessments['precision_assessment']['score']
+        total_similarity_scores = total_similarity_scores + similarity_score
+        total_precision_scores = total_precision_scores + precision_score
+        print(f"eval on {idx}...similarity_score: {similarity_score}, precision_score: {precision_score}")
```
Should be batch this actually on the local LLM-level.

deep-diver replied:
yeah I figured that after making this PR.
since you noticed, I will update the codes!

sayakpaul commented on src/gen/local_lm.py:
```diff
@@ -0,0 +1,47 @@
+
+    return tokenizer, model
+
+def gen_model_output(model, tokenizer, ds, temperature=0.4, max_new_tokens=1024, delimiter="assistant\n"):
```
Here, if we directly pass / allow a list of prompts, it becomes a batched a implementation.

deep-diver replied:
I haven't tried batched inference yet, but is it that simple?
gonna give it a try anyways :)

sayakpaul replied:
Yeah.

sayakpaul reviewed (commented):
Thanks!

Left a couple of comments.

deep-diver committed:
wrap eval results in a dict

deep-diver committed:
rename: ft -> fine-tuning

deep-diver committed:
rename: get_steps_right -> validate_steps

deep-diver committed:
batch evals

deep-diver committed:
fix: wrong idx log

deep-diver committed:
batch processing

deep-diver commented:
@sayakpaul 

addressed comments
: local LLM (gemma)'s batch inference with `dataset.map()` approach.
: makes simultaneous gemini API calls on the batch generated outputs

deep-diver committed:
fix: correctly averaging the output

sayakpaul commented on src/config/sample_config.yaml:
```diff
@@ -0,0 +1,63 @@
+gradient_checkpointing: true
+gradient_checkpointing_kwargs:
+  use_reentrant: false
+hub_model_id: gemma-2b-sft-qlora-no-robots
```
```suggestion
hub_model_id: gemma-7b-sft-qlora-no-robots
```

sayakpaul commented on src/config/sample_config.yaml:
```diff
@@ -0,0 +1,63 @@
+max_seq_length: 2048
+max_steps: -1
+num_train_epochs: 1
+output_dir: data/gemma-2b-sft-qlora-no-robots
```
```suggestion
output_dir: data/gemma-7b-sft-qlora-no-robots
```

sayakpaul commented on src/config/sample_config.yaml:
```diff
@@ -0,0 +1,63 @@
+per_device_train_batch_size: 4
+push_to_hub: true
+report_to:
+- tensorboard
```
```suggestion
- tensorboard
- wandb
```

sayakpaul commented on src/gen/gemini.py:
```diff
@@ -0,0 +1,40 @@
+import google.generativeai as genai
+
+async def call_gemini(prompt="", api_key=None, generation_config=None, safety_settings=None):
+    genai.configure(api_key=api_key)
```
Shouldn't this just be configured only once?

deep-diver replied:
You are right! Thanks for point this out :)

sayakpaul commented on src/gen/gemini.py:
```diff
@@ -0,0 +1,40 @@
+            },
+        ]
+
+    model = genai.GenerativeModel(model_name="gemini-1.0-pro",
+                                generation_config=generation_config,
+                                safety_settings=safety_settings)
```
Same. Shouldn't this be initialized just once?

deep-diver replied:
actually, no.

even if I didn't make CLI options for generation_config, sefety_settings yet, this should be configurable by users.

sayakpaul replied:
Then "gemini-1.0-pro" should be a CLI option too.

deep-diver replied:
right. I don't want this too verbose for right now.
I am thinking about having another config.yaml to setup Gemini (model name, gen config, safe settings, etc.,). But want to handle this in the future PR. What do you think?

sayakpaul replied:
Should be a single change:

`model_name="gemini-1.0-pro"` to `model_name=args.model_name`.

deep-diver replied:
yeah but don't want to put everything into CLI. There are too many parameters for gen_config, safety settings. If I have model_name, I feel like I need to add CLI options for the other two. 

not sure if this project is one time thing, but I am planning to add more service LLM like Claude and ChatGPT. so wanted to manage all relevant settings per service in separate files.

maybe just adding `model_name` CLI option for now, then work on generalization later?

sayakpaul commented on src/gen/local_lm.py:
```diff
@@ -0,0 +1,57 @@
+    quantization_config = BitsAndBytesConfig(load_in_8bit=load_in_8bit, load_in_4bit=load_in_4bit)
+
+    tokenizer = get_tokenizer(model_args, data_args)
+    tokenizer.padding_side = 'left'
```
Should be done within `get_tokenizer()` function itself and should be accompanied with a comment.

deep-diver replied:
I assume you are pointing out `tokenizer.padding_side = 'left'` only here, right?

sayakpaul replied:
Yeah

sayakpaul commented on src/main.py:
```diff
@@ -0,0 +1,78 @@
+import asyncio
+import argparse
+
+from pipeline.eval import eval_on_records
+from utils import validate_steps
```
I feel like `main.py` shouldn't be a part of `src` actually. It should likely be moved out of `src`. So should be `requirements.txt`.

sayakpaul reviewed (commented):
Thanks! Left a couple more comments.

deep-diver committed:
enable gemini api once

deep-diver committed:
make default configs constants

deep-diver committed:
remove unnecessary tokenizer.padding_side setting

deep-diver committed:
leverage sft_args to refer fine-tuned model id

deep-diver committed:
update sample_config.yaml

deep-diver committed:
fix: typo on model_id

deep-diver committed:
move config dir to root

deep-diver committed:
move main, requirements to root

deep-diver committed:
remove gemini model name

deep-diver committed:
add service model name CLI option

deep-diver commented:
@sayakpaul 

addressed your comments!

sayakpaul commented:
Thanks for iterating!
